modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java

```diff
@@ -10,144 +10,153 @@
+                    if (lastGeneration != null) {
+                        throw new NoSuchFileException(
                             "Blob object ["
                                 + blobId.getName()
                                 + "] generation ["
                                 + lastGeneration
                                 + "] unavailable on resume (contents changed, or object deleted): "
                                 + storageException.getMessage()
-                        )
-                    );
-                } else {
-                    throw addSuppressedExceptions(
-                        new NoSuchFileException("Blob object [" + blobId.getName() + "] not found: " + storageException.getMessage())
-                    );
+                        );
+                    } else {
+                        throw new NoSuchFileException("Blob object [" + blobId.getName() + "] not found: " + storageException.getMessage());
+                    }
                 }
-            }
-            if (storageException.getCode() == RestStatus.REQUESTED_RANGE_NOT_SATISFIED.getStatus()) {
-                long currentPosition = Math.addExact(start, currentOffset);
-                throw addSuppressedExceptions(
-                    new RequestedRangeNotSatisfiedException(
+                if (storageException.getCode() == RestStatus.REQUESTED_RANGE_NOT_SATISFIED.getStatus()) {
+                    throw new RequestedRangeNotSatisfiedException(
                         blobId.getName(),
-                        currentPosition,
-                        (end < Long.MAX_VALUE - 1) ? end - currentPosition + 1 : end,
+                        start,
+                        (end < Long.MAX_VALUE - 1) ? end - start + 1 : end,
                         storageException
-                    )
-                );
+                    );
+                }
+                throw storageException;
             }
-            throw addSuppressedExceptions(storageException);
+        }
+
+        @Override
+        public void onRetryStarted(StreamAction action) {
+            // No retry metrics for GCS
+        }
+
+        @Override
+        public void onRetrySucceeded(StreamAction action, long numberOfRetries) {
+            // No retry metrics for GCS
+        }
+
+        @Override
+        public long getMeaningfulProgressSize() {
+            return Math.max(1L, GoogleCloudStorageBlobStore.SDK_DEFAULT_CHUNK_SIZE / 100L);
+        }
+
+        @Override
+        public int getMaxRetries() {
+            return blobStore.getMaxRetries();
+        }
+
+        @Override
+        public String getBlobDescription() {
+            return blobId.toString();
+        }
+
+        @Override
+        public boolean isRetryableException(StreamAction action, Exception e) {
+            return switch (action) {
+                case OPEN -> BaseService.EXCEPTION_HANDLER.shouldRetry(e, null);
+                case READ -> true;
```

Comment:
I think this should be

```suggestion
                case READ -> e instanceof StorageException;
```
since all retryable `IOException` is wrapped with `StorageException.translate`?